File: apps/mcp-server/src/tools/memory.ts
/**
 * MCP tools for the `memory` module.
 *
 * Mirrors `apps/memory/src/mcp/tools.ts`:
 *   - colber_memory_store
 *   - colber_memory_retrieve   (semantic search via `sdk.memory.search`)
 *   - colber_memory_update
 *   - colber_memory_share
 */

import { z } from 'zod';

import type { ToolRegistry } from './registry.js';
import type { ColberClient } from '@colber/sdk';

const MemoryTypeSchema = z.enum(['fact', 'event', 'preference', 'relation']);
const VisibilitySchema = z.enum(['private', 'operator', 'shared', 'public']);

const PermissionsSchema = z.object({
  visibility: VisibilitySchema,
  sharedWith: z.array(z.string().min(1).max(512)).max(256).optional(),
});

const PayloadSchema = z.record(z.unknown());

export const registerMemoryTools = (registry: ToolRegistry, sdk: ColberClient): void => {
  registry.register({
    name: 'colber_memory_store',
    description:
      '[Colber] Persist a new memory (text + structured payload). Generates an embedding via the configured provider. Returns the new memory id.',
    inputSchema: z.object({
      ownerDid: z.string().min(1).max(512),
      type: MemoryTypeSchema,
      text: z.string().min(1).max(64_000),
      payload: PayloadSchema.optional(),
      permissions: PermissionsSchema,
      encryption: z.object({ enabled: z.boolean() }).optional(),
    }),
    handler: (input) =>
      sdk.memory.store({
        ownerDid: input.ownerDid,
        type: input.type,
        text: input.text,
        ...(input.payload !== undefined ? { payload: input.payload } : {}),
        permissions: {
          visibility: input.permissions.visibility,
          ...(input.permissions.sharedWith !== undefined
            ? { sharedWith: input.permissions.sharedWith }
            : {}),
        },
        ...(input.encryption !== undefined ? { encryption: input.encryption } : {}),
      }),
  });

  registry.register({
    name: 'colber_memory_retrieve',
    description:
      '[Colber] Semantic search across memories visible to the caller (top-k by cosine similarity). Permission-aware: only memories the caller is allowed to see can be returned.',
    inputSchema: z.object({
      queryDid: z.string().min(1).max(512),
      queryText: z.string().min(1).max(8_000),
      topK: z.number().int().min(1).max(50).optional(),
      filters: z
        .object({
          type: MemoryTypeSchema.optional(),
          ownerDid: z.string().min(1).max(512).optional(),
          visibility: VisibilitySchema.optional(),
        })
        .optional(),
    }),
    handler: async (input) => {
      const filters =
        input.filters !== undefined
          ? {
              ...(input.filters.type !== undefined ? { type: input.filters.type } : {}),
              ...(input.filters.ownerDid !== undefined ? { ownerDid: input.filters.ownerDid } : {}),
              ...(input.filters.visibility !== undefined
                ? { visibility: input.filters.visibility }
                : {}),
            }
          : undefined;
      return sdk.memory.search({
        queryDid: input.queryDid,
        queryText: input.queryText,
        ...(input.topK !== undefined ? { topK: input.topK } : {}),
        ...(filters !== undefined ? { filters } : {}),
      });
    },
  });

  registry.register({
    name: 'colber_memory_update',
    description:
      "[Colber] Update a memory's text and/or payload. Versioned (audit trail in memory_versions). Re-embeds when text changes. Owner-only.",
    inputSchema: z
      .object({
        id: z.string().uuid(),
        callerDid: z.string().min(1).max(512),
        text: z.string().min(1).max(64_000).optional(),
        payload: PayloadSchema.optional(),
      })
      .superRefine((val, ctx) => {
        if (val.text === undefined && val.payload === undefined) {
          ctx.addIssue({
            code: z.ZodIssueCode.custom,
            message: 'must change at least one of `text` or `payload`',
            path: ['text'],
          });
        }
      }),
    handler: (input) =>
      sdk.memory.update({
        id: input.id,
        callerDid: input.callerDid,
        ...(input.text !== undefined ? { text: input.text } : {}),
        ...(input.payload !== undefined ? { payload: input.payload } : {}),
      }),
  });

  registry.register({
    name: 'colber_memory_share',
    description:
      '[Colber] Grant additional agents read access to a memory. Owner-only. Optional `expiresAt` is recorded for future expiry enforcement.',
    inputSchema: z.object({
      id: z.string().uuid(),
      callerDid: z.string().min(1).max(512),
      shareWith: z.array(z.string().min(1).max(512)).min(1).max(256),
      expiresAt: z.string().datetime().optional(),
    }),
    handler: (input) =>
      sdk.memory.share({
        id: input.id,
        callerDid: input.callerDid,
        shareWith: input.shareWith,
        ...(input.expiresAt !== undefined ? { expiresAt: input.expiresAt } : {}),
      }),
  });
};

File: apps/langchain-colber/src/langchain_colber/tools/memory.py
"""Memory service tools — store, query, share semantic memories."""

from __future__ import annotations

from typing import Any, ClassVar

from pydantic import BaseModel, Field

from ._base import ColberToolBase


class _StoreArgs(BaseModel):
    owner_did: str = Field(
        description="DID of the agent that owns the memory. Becomes the only "
        "principal allowed to update or share it."
    )
    type: str = Field(
        description=(
            "Memory category. One of: ``fact``, ``event``, ``preference``, "
            "``relation``."
        ),
        pattern="^(fact|event|preference|relation)$",
    )
    text: str = Field(
        description="Free-form text to embed and index.",
        min_length=1,
        max_length=64 * 1024,
    )
    visibility: str = Field(
        default="private",
        description=(
            "ACL visibility of the new memory: ``private`` / ``operator`` / "
            "``shared`` / ``public``."
        ),
        pattern="^(private|operator|shared|public)$",
    )
    payload: dict[str, Any] | None = Field(
        default=None,
        description="Optional structured payload stored alongside the text.",
    )


class _QueryArgs(BaseModel):
    query_did: str = Field(
        description=(
            "DID of the agent issuing the query. Only memories owned by — or "
            "shared with — this DID are returned."
        )
    )
    query_text: str = Field(
        description="Natural-language query text. Embedded server-side.",
        min_length=1,
    )
    top_k: int = Field(
        default=5,
        description="Maximum number of hits to return (1..50).",
        ge=1,
        le=50,
    )


class _ShareArgs(BaseModel):
    id: str = Field(description="Memory id returned by ``colber_memory_store``.")
    caller_did: str = Field(
        description=(
            "DID of the agent issuing the share grant — must be the memory "
            "owner."
        )
    )
    share_with: list[str] = Field(
        description="List of DIDs that should gain read access.",
        min_length=1,
    )
    expires_at: str | None = Field(
        default=None,
        description="Optional ISO-8601 expiration timestamp for the grant.",
    )


class MemoryStoreTool(ColberToolBase):
    """Store a new memory in colber-memory."""

    service_name: ClassVar[str] = "memory"

    name: str = "colber_memory_store"
    description: str = (
        "Persist a new semantic memory owned by ``owner_did``. The text is "
        "embedded server-side and indexed for similarity search. Use this "
        "to record facts, events, preferences, or relationships you want to "
        "retrieve later via ``colber_memory_query``."
    )
    args_schema: type[BaseModel] = _StoreArgs

    def _call_colber(self, **kwargs: Any) -> Any:
        return self._client.memory.store(
            owner_did=str(kwargs["owner_did"]),
            type=str(kwargs["type"]),  # type: ignore[arg-type]
            text=str(kwargs["text"]),
            permissions={
                "visibility": str(kwargs.get("visibility", "private")),
                "shared_with": [],
            },
            payload=kwargs.get("payload"),
        )


class MemoryQueryTool(ColberToolBase):
    """Semantic search over memories accessible to ``query_did``."""

    service_name: ClassVar[str] = "memory"

    name: str = "colber_memory_query"
    description: str = (
        "Retrieve up to ``top_k`` memories most relevant to the query text, "
        "scoped to memories ``query_did`` is allowed to see (its own + ones "
        "explicitly shared with it). Returns a list of ``{id, score, type, "
        "owner_did, snippet}`` hits."
    )
    args_schema: type[BaseModel] = _QueryArgs

    def _call_colber(self, **kwargs: Any) -> Any:
        return self._client.memory.search(
            query_did=str(kwargs["query_did"]),
            query_text=str(kwargs["query_text"]),
            top_k=int(kwargs.get("top_k", 5)),
        )


class MemoryShareTool(ColberToolBase):
    """Grant read access on a memory to one or more peer DIDs."""

    service_name: ClassVar[str] = "memory"

    name: str = "colber_memory_share"
    description: str = (
        "Grant read access on a memory record to one or more peer DIDs. The "
        "caller must be the memory's owner. Returns the updated "
        "``shared_with`` list."
    )
    args_schema: type[BaseModel] = _ShareArgs

    def _call_colber(self, **kwargs: Any) -> Any:
        share_with_raw = kwargs["share_with"]
        if isinstance(share_with_raw, list):
            share_with = [str(item) for item in share_with_raw]
        else:
            share_with = [str(share_with_raw)]
        return self._client.memory.share(
            id=str(kwargs["id"]),
            caller_did=str(kwargs["caller_did"]),
            share_with=share_with,
            expires_at=(
                str(kwargs["expires_at"])
                if kwargs.get("expires_at") is not None
                else None
            ),
        )


__all__ = ["MemoryQueryTool", "MemoryShareTool", "MemoryStoreTool"]
